### ISSUE TITLE
Canvas behaves randomly, completely broken

### ISSUE THREAD

**Issue by php4fan** [brave/brave-browser#26530]
## Steps to Reproduce

   1. First visit this page in Chrome or Firefox: https://opentarget.org/test/ssclone2022
   2. now open the same url in Brave

## Expected result:
Should behave the same in Brave as it does in any other browser.

Here's what it should look like after a while:

![image](https://user-images.githubusercontent.com/13969584/199863580-244c2983-13e6-4081-97c1-8d6bb1a258c0.png)

## Actual result:
Notice the pixels falling from the top: they are supposed to keep falling until they hit the "ground", and that's they behave in other browsers.

In Brave, however, at random times some of them remain stuck mid-air. After a few minutes or less, it looks completely broken like this:

![image](https://user-images.githubusercontent.com/13969584/199863622-a7d7f01f-d047-48b4-97df-3e28be8f55d0.png)

(the vertical segments are the result of several pixels falling on top of the ones that got stuck mid-air and accumulating there)

I know this is far from a minimal example isolating the issue, but it's what I have. There's no question that there is a bug.

## Reproduces how often: 
every single time.

## Brave version (brave://version info)

1.45.116 Chromium: 107.0.5304.91 (Official Build) (64-bit) 

## Version/Channel Information:

- Can you reproduce this issue with the current release? yes
- Can you reproduce this issue with the beta channel?  not tried
- Can you reproduce this issue with the nightly channel?  not tried

## Other Additional Information:

- Does the issue resolve itself when disabling Brave Shields? I have no idea, couldn't find how to do that
- Does the issue resolve itself when disabling Brave Rewards? didn't even search
- Is the issue reproducible on the latest version of Chrome? **NO**

## Miscellaneous Information:
Manjaro "brave-browser" package

**Comment by php4fan** [comment]
It seems likely to be this issue (which I can reproduce):
https://community.brave.com/t/canvas-and-uint32array-are-not-working-as-expected/286732

Reported a year ago, no reply and still not fixed. Pathetic.

I don't even want to use Brave, I'm forced to caring about this only because users of my website use it and it breaks it.
I would love to just block Brave by user agent, but unfortunately you disguise it as Chrome 🤦

**Comment by rebron** [comment]
@php4fan In shields, if you select `Allow fingerprinting` does that work for you. I'm not able to reproduce on Windows and macOS, I see what you're seeing with the Expected Result image you posted so maybe a Linux only issue.

**Comment by php4fan** [comment]
> I'm not able to reproduce on Windows and macOS

Just with my test case or also the blank-canvas one I linked in my last comment?

> if you select Allow fingerprinting does that work for you

By that I guess you mean selecting "Disabled" under "Block fingerprinting":

![image](https://user-images.githubusercontent.com/13969584/200028835-595ed302-6dc9-4d0e-bc9c-a03702ca7b9b.png)

**Unbelievably, that does work.** It fixes both my falling-pixels example and the blank-canvas test from StackOverflow.

```
Disabled => systematically works
Standard => systematically broken
Strict   => haven't tested
```
(or almost systematically, as I have only tried switching half a dozen times)

I'm really curious to know how something like blocking fingerprinting can possibly break reading pixel color values from a canvas 🤦

> so maybe a Linux only issue.

Could be... The user I talked to who found the issue in my example was also using linux. However, the one in the community forum thread mentions the blank-canvas test breaking on MacOS and Windows, though that post is from a year ago.

**Comment by rebron** [comment]
cc: @pilgrim-brave

**Comment by php4fan** [comment]
>  WHY??

Ok so I [read](https://brave.com/privacy-updates/3-fingerprint-randomization/) _ [about](https://browserleaks.com/canvas) the why and it's pure madness. This "randomization" approach to fingerprinting prevention is just unbelievably stupid. It's great that you explore crazy ideas (it's called "Brave" after all), but this particular idea should have been discarded after thinking about it for five minutes, let alone testing it and even seeing it in production for months. If "privacy" (and I put it in quotes because we're talking about fingerprinting prevention) is more important to you than **a programming language working to specifications**, then I'm sorry, but the only way to be 100% protected is to not use the internet. Make a browser that just opens a blank window and doesn't make any http request: that will be the fastest, most lightweight browser ever, and it will be 100% fingerprinting-proof.

Let's leave aside for a moment how laughable it is that you pride yourselves of using randomization _"instead of [...] adding troublesome permission prompts"_ (that is, instead of doing the rational, sensible thing). Let's pretend that this randomization approach is not ridiculous in and of itself.

You need, at the very least, to **move the "add random noise to canvas" thing from the _Standard_ to the _Extreme_** setting (or _Strict_ or whatever it's called, the one that is not the default, the one that si expected to break stuff and is meant for the paranoids who care for fingerprinting prevention more than they care for their lives). 

It **cannot be the default** to have the following assertion break in a programming language:
- write an integer `n` into an array at index `i`
- read the value in the array at index `i`, let it be `m` 
- `m == n` is guaranteed to be true

And I hate to break it to you, but a canvas is just that at a fundamental level: an array of 32-bit integers.

Let me just break down the list of assumptions that must hold in order for it to be acceptable to add random noise to a canvas (which, remember, means adding random `+1`s and `-1`s to integer values stored in an array, let that sink in):
- a canvas is just a space for drawing _**pictures**_ 
- where _pictures_ means the kind of images where **differences that are not perceptible to the human eye don't matter**
- **no kind of non-linear image processing** of an image (which may cause non-perceptible differences to become dramatic differences) is conceivable or remotely likely in a real-life use case of canvases
- **reading _exact_ pixel color values out of an image** has no remotely likely real-life use cases

These assumptions are **_extremely_ narrow-minded**. And arrogant. Yes, _arrogant_. Because they assume that if you haven't seen a use case or been able to think of one, it doesn't exist. Innovation happens thanks to people coming up with clever use cases that nobody has thought of before (ironically, just like when someone comes up with using randomization to prevent fingerprinting - but I'm talking about clever ideas that turn out to make sense). By assuming that there's only a finite set of legitimate, sensible, useful uses of any given tool (like a canvas) which is the set of use cases that are currently commonly seen, **means preventing innovation**. That's not _brave_ at all.

But that's just a consideration about the general mindset that is behind this kind of design decisions. In this particular case, it's way more basic than that: you are **breaking plenty of obvious, basic use cases that already exist**.

**Comment by SlugFiller** [comment]
I just want to point out that the randomization solution is also counter-productive. What it tries to prevent is a case of browsers using different details in the rendering, allowing the specific algorithms used by your browser to be identified, therefore identifying the browser.

There is no browser other than Brave doing "canvas randomization". The simple check of clearing the canvas (identical result on 99% of the browsers), then checksumming it, will instantly identify Brave. That is the opposite of what "anti fingerprinting" is supposed to do. Repeating it a few times can fingerprint the RNG too.

The real solution is to use highly deterministic/reproducible algorithms, so architecture details, like the type of floating point operations used in the CPU/GPU, cannot be detected. It would be a performance hit, but would guarantee all Brave users appear the same, and can possibly pass off as another browser as well.

**Comment by php4fan** [comment]
> I just want to point out that the randomization solution is also **counter-productive**
> What it tries to prevent is [...] **identifying the browser**.

Well to be fair (although obviously I too am against Brave's randomization approach because it is insane), the fact that it allows to identify Brave itself is not, I believe, a concern.

1) one (meaning whoever wants to fingerprint you) can already use `navigator.brave` to tell whether or not you are using brave (and luckily so, because it's what allows me to make my website, which cannot work properly because of this stupid bug, refuse to run in Brave).
2) Identifying the broad "brand" of the browser (i.e. Chrome vs Brave vs Edge vs Firefox etc) I don't think is considered a big deal in the context of fingerprinting, in that it doesn't "narrow you down" a lot. Also I guess it's considered almost unavoidable. What one would want to prevent is for the specific version to be identified, as well as things like the CPU, operating system, user settings, etc.
3) even more so if one assumes that Brave is amazing (i.e. in a hypothetical future) and therefore becomes very widely adopted: then identifying that you are using Braves would become even less meaningful.

So I agree that it is counter-productive, but **not in that it defeats the purpose of fingerprinting-prevention**. But rather, because it defeats the purpose of the browser itself: which is to browse websites. It breaks everything and turns the WWW into an unusable, unpredictable mess, by having JavaScript behave in a non-deterministic and non-standard-compliant way. And it will eventually go against Brave adoption, because it will cause content authors to do the same thing that I did: prevent their content from running in Brave; which will drive away users even more than broken websites does by itself.

**Comment by SlugFiller** [comment]
@php4fan Identifying the broad "brand" of the browser doesn't "narrow you down" a lot, yes, but it is exactly the level of "narrowing down" that canvas fingerprinting originally achieves. As the articles explain, this is combined with a large amount of other data to farther narrow it down. This is, in fact, the main output the browser leak demos show - An estimation of which browser you're using. They also explain some of the "mis-readings" in their signature database are coming from users faking their user agents, which is what canvas and audio based fingerprinting allows them to defeat.

The point isn't you can find the same with UA analysis, or looking for `navigator.brave`. It's that those two things should be the first things one would expect anti-fingerprinting to fake and hide. If it's not, the whole feature might be pointless. Like barricading the door and leaving the window open.

P.S. You shouldn't test for `navigator.brave` either. Take a page from the JQuery code - to check for various browser defects (*cough* Internet Explorer *cough*), they feed broken HTML string into a div's innerHTML, then read back the elements to examine how it was interpreted. Likewise, you can run a test where you clear a 2x2 canvas, and read back the data, repeat a few times, and if it doesn't come back clear, you detect a "Brave-exclusive standard-breaking behavior", and treat it as you would treat a detection of IE5.

P.P.S. The even bigger issue is that it doesn't just detect Brave, but actually detects whether you have anti-fingerprinting on or off. You know how many web pages have a "stop your ad blocker to continue" popup if you browse them in a private tab? Imagine if they had a non-skippable version if you browse them in anti-fingerprinting mode, which gives instructions on how to turn it off, and doesn't let you proceed any farther until you do. To me, what anti-fingerprinting SHOULD be doing, is making that popup not show up even for private browsing, i.e. making it harder for websites to detect my exact browser settings.

**Comment by php4fan** [comment]
> @php4fan Identifying the broad "brand" of the browser doesn't "narrow you down" a lot, yes, but it is exactly the level of "narrowing down" that canvas fingerprinting originally achieves

Really? I thought it identified way more than that. The way I understand it, it consists of things like rendering text into the canvas and checking the result: that would for example reveal what fonts are being used by default which in turn tell you stuff about the operating system and the like. And then I imagine things like analyzing the results of interpolating images, which would tell you stuff about the GPU maybe.

If it's like you say and it only serves the purpose of identifying the browser "brand", then yes, canvas randomisation, besides being a ridiculous approach that breaks everything and turns the browser into an unpredictable mess that spits out random garbage, it is *also* basically pointless 🤦‍♂️

Anyway, it seems unlikely that they'll ever give up this idiotic randomisation approach as a whole, given how proud they are of their brilliant idea and how much they have invested on it (unless they are smart enough to recognize their mistakes, but if they were smart enough they would have seen it after 5 minutes, not several months); but even accepting the insane premise of using randomisation in the first place, I hope they understand that at the very least they need to move the canvas madness from the default setting to the "be-aware-this-breaks-everything" one.

**Comment by php4fan** [comment]
I see, I guess you're right then. All the more reason to remove this idiocy, since is is not even useful for what it's supposed to achieve 🤦‍♂️

**Comment by pes10k** [comment]
Im closing this out as non-actionable.  But will note for anyone who finds this that

> What I do get is that it seems to retain the same incredibly unique signature until I restart the browser

is intentional behavior. We align the browser fingerprint with the browser session. Sites can already reidentify you w/in the same browser session using DOM storage. The purpose of brave's protections are to give you a new fingeprint per site, per session. 

> The default Chromium behavior is better and more secure

This is extremely incorrect

**Comment by php4fan** [comment]
> Im closing this out as non-actionable. 

Are you stupid???

This is extremely easily actionable. Canvas randomization should either be completely removed (as it breaks everything and converts the browser into a shitshow of complete randomness that doesn't comply with the most basic standards), or only enabled when the setting for fingerprinting protection is set to **"strict, may break sites"** and not "standard".

Have you read https://github.com/brave/brave-browser/issues/26530#issuecomment-1304500812

What is your response to that, to justify not fixing this?

**Comment by php4fan** [comment]
Here's a list of examples of real-life applications that could all potentially miserably break because of this bug. Not only that, but they would behave in a completely random and unexpected fashion with no clue for the user that the cause has anything to do with fingerprinting, or that any of the completely random observed behavior is unexpected in the first place.

I intended to implement each of these as a demo but I haven't had enough free time to throw to the garbage.

1. A [Conway's Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) simulation where the grid is stored and at the same time visualized using a canvas (scaled up), and each cell in the grid is one canvas pixel, whose status is represented by the pixel being either black or white
2. A demonstration of a **chroma key** effect. Actually, I have found an exaple in the wild that breaks in Brave: http://examples.hmp.is.it/chromakeycanvas.php. To make it evident, set the tolerance to zero and the color to `#FF0000`. Watch the difference between Brave and any other browser.

Well, I had a few more but I am too lazy to go on. Just one example should be enough. This behavior breaks websites and therefore it should not exist in the "standard" mode (if at all).

**Comment by pes10k** [comment]
We're blocking a user from this thread. We greatly appreciate input and bug reports, but coming into where we work and throwing insults, using hostile language, and other similar unpleasantness is not welcome or tolerated.

For anyone else coming across this thread, brave's canvas fingeprinting protections in canvas are based around the plan observation that the feautre is misused for fingerprinting far (far far) more than its used for user serving purposes. We only randomize on canvas readback, not when writing to the canvas, so most canvas use is also unaffected.  Finally, we only randomize slightly, so most canvas readback uses are also not affected.

There are still a small number of remaining uses that we'd like to support when shields are up, and we're experimenting with other approaches that might do so.  But, in the meantime, the current defense is a best effort to balance on one hand frustrating frequent fingerprinting efforts, while on the other hand enabling as many benign / user-desirable uses as possible. We can keep improving, but the current choice is thought through and we're aware of the trade offs. Insults definitely are not needed or helpful

**Comment by rofoto** [comment]
This issue completely breaks performance on any site that uses canvas and needs constant user feedback(games and other interactive applications) which I would argue is most use cases of canvas.

 My own interactive projects are totally unusable in brave as they require user input; either camera based interaction which the user is knowingly agreeing to or mouse inputs that are read as fast as the canvas can update. Also attempting to do simulations where forces are calculated in sub 0.00x values basically break if the browser is inserting whole digit ints across fields.

So basically any simulations that are meant to be in real time ( ie 30/60 fps ) are not able to be run in Brave. 

Giving the user the option to turn this off when encountering any canvas element; while not ideal from a UI/UX perspective would at least give the visitor the option to turn this off for at least the length of their visit to this specific site/tab. 

Having it buried in settings just makes it look like the sites are bad, not that the browser is intentionally hurting performance.

**Comment by pes10k** [comment]
@rofoto , thanks for the feedback.  This isn't quite correct though.

We only add a small amount of randomization to the canvas on readback.  If you're just drawing to canvas, we don't change anything.  So, as long as you're not reading things back into a buffer before drawing them to a canvas element, you should't be seeing any modifications.

There is also a much easier way for your users to disable these protections for your site, if they trust your site to do so.  They can control shields per site, w/o going into settings.  Hit the lion / shields dialog in the address bar, hit advanced controls, and hit allow fingerprinting.  That'll disable this fingerprinting protection for your domain.
<img width="389" alt="Screenshot 2023-10-25 at 10 06 39" src="https://github.com/brave/brave-browser/assets/76526/bd123d2b-b1c8-4eaf-94ad-3bee56fc419a">

**Comment by rofoto** [comment]
> ...you're not reading things back into a buffer before drawing them to a canvas element,...

this is exactly what I am doing. And sadly there is no real way around it

> There is also a much easier way for your users to disable these protections for your site, if they trust your site to do so....

basically this is a non starter, From a UX perspective having to display a how disable something to make this work for a website like this would just have that user not go to the site plus it adds a sketchy vibe to the entire experience.  

its far easier to just either detect brave (obviously this is not as easy since that is the entire point of brave AFAIK)
and refuse to start with a list of browser options that are known to work.

Or just flat out say 'this project will perform badly on brave' and recommend they use a more compatible browser

However if the browser itself pops up a message; which brave has done for me on some sites as I do use it as my main browser,  pointing out that the site will not perform optimally unless they disable shields then its better integrated with the whole use of the browser. If there is a js api call that can trigger that pop up from my side then sure I could use that but outside of that its just easier for the user to just say don't use brave.

**Comment by pes10k** [comment]
Understood.  The JS API approach is appealing, but not something we have bandwidth for in immediate term.  Thank you for the suggestion, and we'll follow up if / when we make any changes in this regard
